src/parser/jobs/sam/modules/Shoha.tsx

```diff
@@ -0,0 +1,272 @@
+	static displayMode = DISPLAY_MODE.FULL
+
+	private stacks = 0
+	private shohaUses = 0
+	private stackHistory: StackState[] = []
+	private wasteBySource = {
+		[ACTIONS.HIGANBANA.id]: 0,
+		[ACTIONS.KAESHI_HIGANBANA.id]: 0,
+		[ACTIONS.TENKA_GOKEN.id]: 0,
+		[ACTIONS.KAESHI_GOKEN.id]: 0,
+		[ACTIONS.MIDARE_SETSUGEKKA.id]: 0,
+		[ACTIONS.KAESHI_SETSUGEKKA.id]: 0,
+		[ACTIONS.RAISE.id]: 0,
+	}
+	private leftoverStacks = 0
+	private totalGeneratedStacks = 0 // Keep track of the total amount of generated stacks over the fight
+
+	@dependency private checklist!: Checklist
+	@dependency private suggestions!: Suggestions
+	@dependency private combatants!: Combatants
+
+	protected init() {
+		this.addHook('init', this.pushToHistory)
+		this.addHook(
+			'cast',
+			{
+				by: 'player',
+				abilityId: Object.keys(GENERATORS).map(Number),
+			},
+			this.onGenerator,
+		)
+		this.addHook(
+			'cast',
+			{
+				by: 'player',
+				abilityId: Object.keys(SPENDERS).map(Number),
+			},
+			this.onSpender,
+		)
+		this.addEventHook([
+			'applybuff', 'applybuffstack'],
+			{
+			to: 'player',
+			abilityId: STATUSES.MEDITATION.id,
+			}, this.onMeditateGain)
+
+		this.addHook('death', {to: 'player'}, this.onDeath)
+		this.addHook('complete', this.onComplete)
+	}
+
+	private onGenerator(event: CastEvent) {
+		if (this.stacks === MAX_STACKS) {
+		const abilityId = event.ability.guid
+		const generatedStacks = GENERATORS[abilityId]
+
+		this.addGeneratedStackAndPush(generatedStacks, abilityId)
+	}
+	}
+
+		// This is exclusively for stacks gained via Meditate channel. and not any stacks gained from GCD casts
+	private onMeditateGain() {
+			this.stacks++
+			this.totalGeneratedStacks++
+
+			// just a safety net
+			if (this.stacks > MAX_STACKS) {
+				this.stacks = MAX_STACKS
+			}
+
+			this.pushToHistory()
+	}
+
+	private addGeneratedStackAndPush(generatedStacks: number, abilityId: number) {
+		this.stacks += generatedStacks
+		this.totalGeneratedStacks += generatedStacks
+		if (this.stacks > MAX_STACKS) {
+			const waste = this.stacks - MAX_STACKS
+			this.wasteBySource[abilityId] += waste
+			this.stacks = MAX_STACKS
+		}
+
+		this.pushToHistory()
+	}
+
+	private onSpender(event: CastEvent) {
+		this.stacks = this.stacks - SPENDERS[event.ability.guid]
+		this.shohaUses++
+
+		// safety net!
+
+		if (this.stacks < 0) {
+			this.stacks = 0
+		}
+
+		this.pushToHistory()
+
+	}
+
+	private onDeath() {
+		this.wasteBySource[ACTIONS.RAISE.id] += this.stacks
+		this.dumpRemainingResources()
+	}
+
+	private dumpRemainingResources() {
+		this.stacks = 0
+		this.pushToHistory()
+	}
+
+	private pushToHistory() {
+		const timestamp = this.parser.currentTimestamp - this.parser.fight.start_time
+		this.stackHistory.push({t: timestamp, y: this.stacks})
+	}
+
+	private onComplete() {
+		this.dumpRemainingResources()
+
+		// final use amounts
+
+		const totalPossibleUses = Math.floor(this.totalGeneratedStacks/MAX_STACKS)
+		const totalUses = this.shohaUses
```

Comment:
Kinda useless assignment, the total possible too. You're not using them in the text or more than just the one value/target, so may as well just have the direct values there. Maybe useful for future additions tho.

Reply:
I felt that seeing 43/44 meditation stacks wouldn't be as useful of a metric as the punishment from overcapping only comes from when you lose a 3n amount of stacks.

IE: 15/16 shoha casts means you missed 3 stacks, where as a 43/44 is just saying you had 2 extra stacks, 1 of which you wasted by overcapping it. which is literally a 0 dps loss

I could just directly use the shohaUses though for this.